modules/ip/datapath/ip_fragment.c

```diff
@@ -0,0 +1,188 @@
+
+		// Check if packet is already a fragment - if so, just pass it through
+		if (ip->fragment_offset
+		    & RTE_BE16(RTE_IPV4_HDR_MF_FLAG | RTE_IPV4_HDR_OFFSET_MASK)) {
+			// This is already a fragment, drop it
+			edge = ALREADY_FRAGMENTED;
+			goto drop;
+		}
+
+		iface = mbuf_data(mbuf)->iface;
+		assert(iface != NULL);
+
+		ip_hdr_len = rte_ipv4_hdr_len(ip);
+		data_len = rte_be_to_cpu_16(ip->total_length) - ip_hdr_len;
+
+		// Calculate fragment payload size (multiple of 8, >= 8)
+		uint16_t max_payload = (uint16_t)(iface->mtu - ip_hdr_len);
+		frag_size = RTE_ALIGN_FLOOR(max_payload, 8);
+		if (unlikely(frag_size < 8)) {
+			edge = ERROR;
+			goto error;
+		}
+
+		num_frags = (data_len + frag_size - 1) / frag_size;
+		assert(num_frags > 1);
+
+		// Create the next fragments
+		for (i = 1; i < num_frags; i++) {
+			// Create new fragment, copying the original IPv4 header.
+			frag_mbuf = rte_pktmbuf_copy(mbuf, mbuf->pool, 0, ip_hdr_len);
+			if (unlikely(frag_mbuf == NULL)) {
+				edge = NO_MBUF;
+				goto error;
+			}
+
+			frag_ip = rte_pktmbuf_mtod(frag_mbuf, struct rte_ipv4_hdr *);
+			offset = i * frag_size;
+			frag_data_len = RTE_MIN(frag_size, data_len - offset);
+
+			payload = rte_pktmbuf_append(frag_mbuf, frag_data_len);
+			if (unlikely(payload == NULL)) {
+				rte_pktmbuf_free(frag_mbuf);
+				edge = ERROR;
+				goto error;
+			}
+
+			memcpy(payload,
+			       rte_pktmbuf_mtod_offset(mbuf, const void *, ip_hdr_len + offset),
+			       frag_data_len);
+
+			frag_ip->total_length = rte_cpu_to_be_16(ip_hdr_len + frag_data_len);
+			frag_ip->fragment_offset = rte_cpu_to_be_16(
+				(offset / 8) | ((i < num_frags - 1) ? RTE_IPV4_HDR_MF_FLAG : 0)
+			);
+			frag_ip->hdr_checksum = 0;
+			frag_ip->hdr_checksum = rte_ipv4_cksum(frag_ip);
+
+			*ip_output_mbuf_data(frag_mbuf) = *ip_output_mbuf_data(mbuf);
+			frag_mbuf->packet_type = mbuf->packet_type;
+			if (gr_mbuf_is_traced(mbuf)) {
+				struct ip_fragment_trace_data *t;
+				t = gr_mbuf_trace_add(frag_mbuf, node, sizeof(*t));
+				t->packet_id = rte_be_to_cpu_16(frag_ip->packet_id);
+				t->frag_num = i;
+				t->offset = offset;
+				t->more_frags = (i < num_frags - 1) ? 1 : 0;
+			}
+
+			rte_node_enqueue_x1(graph, node, IP_OUTPUT, frag_mbuf);
+		}
+
+		// Use original mbuf for first fragment
+		ip->total_length = rte_cpu_to_be_16(ip_hdr_len + frag_size);
+		ip->fragment_offset = RTE_BE16(RTE_IPV4_HDR_MF_FLAG);
+		ip->hdr_checksum = 0;
+		ip->hdr_checksum = rte_ipv4_cksum(ip);
+
+		// Trim excess payload
+		if (unlikely(rte_pktmbuf_trim(mbuf, data_len - frag_size) < 0)) {
+			edge = ERROR;
+			goto error;
+		}
+
+		if (gr_mbuf_is_traced(mbuf)) {
+			struct ip_fragment_trace_data *t;
+			t = gr_mbuf_trace_add(mbuf, node, sizeof(*t));
+			t->packet_id = rte_be_to_cpu_16(ip->packet_id);
+			t->frag_num = 0;
+			t->offset = 0;
+			t->more_frags = 1;
+		}
+
+		edge = IP_OUTPUT;
+
+drop:
+		rte_node_enqueue_x1(graph, node, edge, mbuf);
+		continue;
```

Comment:
This must be done **before** entering the `for (i = 1; i < num_frags; i++)` loop. Otherwise the first fragment will be sent at the end.